panic: runtime error: index out of range

Seen just now in CI, haven't investigated yet:

```
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/golang/dep/internal/gps.(*solver).findValidVersion(0xc42020c6c0, 0xc4205220a0, 0xc420211090, 0x1, 0x1, 0x0, 0xc420211090)
	/home/ubuntu/.go_workspace/src/github.com/golang/dep/internal/gps/solver.go:823 +0x562
github.com/golang/dep/internal/gps.(*solver).createVersionQueue(0xc42020c6c0, 0xc4202a8d40, 0x10, 0x0, 0x0, 0xc420211090, 0x1, 0x1, 0x0, 0x0, ...)
	/home/ubuntu/.go_workspace/src/github.com/golang/dep/internal/gps/solver.go:780 +0xafc
github.com/golang/dep/internal/gps.(*solver).solve(0xc42020c6c0, 0x0, 0x0, 0x2a)
	/home/ubuntu/.go_workspace/src/github.com/golang/dep/internal/gps/solver.go:397 +0x471
github.com/golang/dep/internal/gps.(*solver).Solve(0xc42020c6c0, 0x3b, 0xa760c0, 0xac4ab8, 0xc4200180e3)
	/home/ubuntu/.go_workspace/src/github.com/golang/dep/internal/gps/solver.go:348 +0xab
main.(*ensureCommand).Run(0xc420127960, 0xc420129880, 0xc42000c4a0, 0x0, 0x0, 0x0, 0x0)
	/home/ubuntu/.go_workspace/src/github.com/golang/dep/cmd/dep/ensure.go:149 +0x3c8
main.(*Config).Run(0xc42001ef60, 0xc42001ef60)
	/home/ubuntu/.go_workspace/src/github.com/golang/dep/cmd/dep/main.go:159 +0x8b3
main.main()
	/home/ubuntu/.go_workspace/src/github.com/golang/dep/cmd/dep/main.go:44 +0x253
```

https://circleci.com/gh/weaveworks/common/132

It does indeed seem to be something wrong with golang.org/x/sys. With verbose ensuring:
```
(5)	? attempt golang.org/x/sys with 1 pkgs; at least 1 versions to try
(5)	    try golang.org/x/sys@master
(5)	✗   Unable to update checked out version: fatal: reference is not a tree: a55a76086885b80f79961eacb876ebd8caf3868d
(5)	    try golang.org/x/sys@master
(5)	✗   Unable to update checked out version: fatal: reference is not a tree: a55a76086885b80f79961eacb876ebd8caf3868d
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/golang/dep/internal/gps.(*solver).findValidVersion(0xc420214c60, 0xc4203f9ae0, 0xc4201fa570, 0x1, 0x1, 0x0, 0xc4201fa570)
	/home/ubuntu/.go_workspace/src/github.com/golang/dep/internal/gps/solver.go:823 +0x562
github.com/golang/dep/internal/gps.(*solver).createVersionQueue(0xc420214c60, 0xc420128600, 0x10, 0x0, 0x0, 0xc4201fa570, 0x1, 0x1, 0x0, 0x0, ...)
	/home/ubuntu/.go_workspace/src/github.com/golang/dep/internal/gps/solver.go:780 +0xafc
github.com/golang/dep/internal/gps.(*solver).solve(0xc420214c60, 0x0, 0x0, 0x2a)
	/home/ubuntu/.go_workspace/src/github.com/golang/dep/internal/gps/solver.go:397 +0x471
github.com/golang/dep/internal/gps.(*solver).Solve(0xc420214c60, 0x3b, 0xa760c0, 0xac4ab8, 0xc420018163)
	/home/ubuntu/.go_workspace/src/github.com/golang/dep/internal/gps/solver.go:348 +0xab
main.(*ensureCommand).Run(0xc420127940, 0xc4201299c0, 0xc420010530, 0x0, 0x0, 0x0, 0x0)
	/home/ubuntu/.go_workspace/src/github.com/golang/dep/cmd/dep/ensure.go:149 +0x3c8
main.(*Config).Run(0xc42001afc0, 0xc42001afc0)
	/home/ubuntu/.go_workspace/src/github.com/golang/dep/cmd/dep/main.go:159 +0x8b3
main.main()
	/home/ubuntu/.go_workspace/src/github.com/golang/dep/cmd/dep/main.go:44 +0x253

go get -u github.com/golang/dep/cmd/dep && cd ${REPO_ROOT}/${CIRCLE_PROJECT_REPONAME} && dep ensure -v returned exit code 2

Action failed: go get -u github.com/golang/dep/cmd/dep && cd ${REPO_ROOT}/${CIRCLE_PROJECT_REPONAME} && dep ensure -v
```

--- a/internal/gps/source.go
+++ b/internal/gps/source.go
@@ -246,9 +246,25 @@ func (sg *sourceGateway) exportVersionTo(ctx context.Context, v Version, to stri
 		return err
 	}
 
-	return sg.suprvsr.do(ctx, sg.src.upstreamURL(), ctExportTree, func(ctx context.Context) error {
+	err = sg.suprvsr.do(ctx, sg.src.upstreamURL(), ctExportTree, func(ctx context.Context) error {
 		return sg.src.exportRevisionTo(ctx, r, to)
 	})
+
+	// It's possible (in git) that we may have tried this against a version that
+	// doesn't exist in the repository cache, even though we know it exists in
+	// the upstream. If it looks like that might be the case, update the local
+	// and retry.
+	// TODO(sdboyer) It'd be better if we could check the error to see if this
+	// actually was the cause of the problem.
+	if err != nil && sg.srcState&sourceHasLatestLocally == 0 {
+		if _, err = sg.require(ctx, sourceHasLatestLocally); err != nil {
+			err = sg.suprvsr.do(ctx, sg.src.upstreamURL(), ctExportTree, func(ctx context.Context) error {
+				return sg.src.exportRevisionTo(ctx, r, to)
+			})
+		}
+	}
+
+	return err
 }
 
 func (sg *sourceGateway) getManifestAndLock(ctx context.Context, pr ProjectRoot, v Version, an ProjectAnalyzer) (Manifest, Lock, error) {
@@ -276,6 +292,27 @@ func (sg *sourceGateway) getManifestAndLock(ctx context.Context, pr ProjectRoot,
 		m, l, err = sg.src.getManifestAndLock(ctx, pr, r, an)
 		return err
 	})
+
+	// It's possible (in git) that we may have tried this against a version that
+	// doesn't exist in the repository cache, even though we know it exists in
+	// the upstream. If it looks like that might be the case, update the local
+	// and retry.
+	// TODO(sdboyer) It'd be better if we could check the error to see if this
+	// actually was the cause of the problem.
+	if err != nil && sg.srcState&sourceHasLatestLocally == 0 {
+		// TODO(sdboyer) we should warn/log/something in adaptive recovery
+		// situations like this
+		_, err = sg.require(ctx, sourceHasLatestLocally)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		err = sg.suprvsr.do(ctx, label, ctGetManifestAndLock, func(ctx context.Context) error {
+			m, l, err = sg.src.getManifestAndLock(ctx, pr, r, an)
+			return err
+		})
+	}
+
 	if err != nil {
 		return nil, nil, err
 	}
@@ -310,6 +347,27 @@ func (sg *sourceGateway) listPackages(ctx context.Context, pr ProjectRoot, v Ver
 		ptree, err = sg.src.listPackages(ctx, pr, r)
 		return err
 	})
+
+	// It's possible (in git) that we may have tried this against a version that
+	// doesn't exist in the repository cache, even though we know it exists in
+	// the upstream. If it looks like that might be the case, update the local
+	// and retry.
+	// TODO(sdboyer) It'd be better if we could check the error to see if this
+	// actually was the cause of the problem.
+	if err != nil && sg.srcState&sourceHasLatestLocally == 0 {
+		// TODO(sdboyer) we should warn/log/something in adaptive recovery
+		// situations like this
+		_, err = sg.require(ctx, sourceHasLatestLocally)
+		if err != nil {
+			return pkgtree.PackageTree{}, err
+		}
+
+		err = sg.suprvsr.do(ctx, label, ctGetManifestAndLock, func(ctx context.Context) error {
+			ptree, err = sg.src.listPackages(ctx, pr, r)
+			return err
+		})
+	}
+
 	if err != nil {
 		return pkgtree.PackageTree{}, err
 	}

